### PR TITLE
feat(x/ecocredit): Msg Server implementation for buy/sell functionality

### PR DESCRIPTION
**Pull request by clevinson**
## Description

Closes: #603 #614 #622

This pull request implements the following:

- proto type definitions for `SellOrder`, `BuyOrder`, and `AskDenom`
- proto event definitions for `EventSell`, `EventUpdateSellOrder`, `EventBuyOrderCreated`, `EventBuyOrderFilled`, and `EventAllowAskDenom`
- message service methods for `Sell`, `UpdateSellOrders`, `Buy`, and `AllowAskDenom`
- `ValidateBasic` for `MsgSell`, `MsgUpdateSellOrders`, `MsgBuy`, and `MsgAllowAskDenom`

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/server/testsuite/suite.go [hunk 1]
```diff
@@ -973,6 +973,14 @@ func (s *IntegrationTestSuite) TestScenario() {
 		})
 	}
 
+	// TODO: test create sell order
+
+	// TODO: test update sell order
+
+	// TODO: test create buy order
+
+	// TODO: test allow ask denom
```
Comment (technicallyty) [inline x/ecocredit/server/testsuite/suite.go:982]: will this be another PR or will this PR update?

**Review by technicallyty — CHANGES_REQUESTED** [review]

File: x/ecocredit/msgs.go [hunk 2]
```diff
@@ -375,3 +377,156 @@ func (m *MsgUpdateClassMetadata) GetSigners() []sdk.AccAddress {
...
+
+	for i := range m.Orders {
+		if err := ValidateDenom(m.Orders[i].BatchDenom); err != nil {
+			return err
+		}
+
+		_, err := math.NewPositiveDecFromString(m.Orders[i].Quantity)
+		if err != nil {
+			return sdkerrors.Wrapf(err, "quantity must be positive decimal: %s", m.Orders[i].Quantity)
+		}
```
Comment (aleem1314) [inline x/ecocredit/msgs.go:407]:
```suggestion
		if _, err := math.NewPositiveDecFromString(m.Orders[i].Quantity); err!= nil {
			return sdkerrors.Wrapf(err, "quantity must be positive decimal: %s", m.Orders[i].Quantity)
		}
```

File: proto/regen/ecocredit/v1alpha1/events.proto [hunk 3]
```diff
@@ -103,3 +105,95 @@ message EventCancel {
...
+
+  // new_ask_price is the new ask price for this sell order
+  cosmos.base.v1beta1.Coin new_ask_price = 5;
+
+  // disable_auto_retire updates the disable_auto_retire field in the sell order.
+  bool disable_auto_retire = 6;
+}
+
+// EventBuy is an event emitted when a buy order is created.
+message EventBuy {
```
Comment (aaronc) [inline proto/regen/ecocredit/v1alpha1/events.proto:156]: Should we have separate events when buy orders are created and filled?

Comment (ryanchristo) [reply proto/regen/ecocredit/v1alpha1/events.proto:156]: How would this work? We're not storing buy orders in state, so the created event would emit if partial or filled and the filled event would emit only if the buy order was filled?

File: x/ecocredit/server/msg_server.go [hunk 4]
```diff
@@ -577,3 +516,346 @@ func (s serverImpl) isCreatorAllowListed(ctx types.Context, allowlist []string,
...
+	for i := range req.Updates {
+
+		sellOrder, err := s.getSellOrder(ctx, req.Updates[i].SellOrderId)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO: Verify that NewAskPrice.Denom is in AllowAskDenom #624
+
+		err = verifyBalance(store, ownerAddr, sellOrder.BatchDenom, req.Updates[i].NewQuantity)
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply x/ecocredit/server/msg_server.go]: quantity must be a positive decimal to pass validate basic

[the remark this answers is not in the document]
Comment (ryanchristo) [reply x/ecocredit/server/msg_server.go]: `NewQuantity` must be a positive decimal to pass validate basic

File: x/ecocredit/server/msg_server.go [hunk 5]
```diff
@@ -577,3 +516,346 @@ func (s serverImpl) isCreatorAllowListed(ctx types.Context, allowlist []string,
...
+		if err != nil {
+			return nil, err
+		}
+
+		sellOrder.Quantity = req.Updates[i].NewQuantity
+		sellOrder.AskPrice = req.Updates[i].NewAskPrice
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply x/ecocredit/server/msg_server.go]: `NewAskPrice` must not be nil to pass validaate basic

File: x/ecocredit/server/testsuite/suite.go (hunk 1 above)
Comment (ryanchristo) [reply x/ecocredit/server/testsuite/suite.go]: https://github.com/regen-network/regen-ledger/issues/627

File: proto/regen/ecocredit/v1alpha1/events.proto (hunk 3 above)
Comment (technicallyty) [reply proto/regen/ecocredit/v1alpha1/events.proto:156]: i think this would work for orders with filters, where buy orders can be made for a credit with X characteristics, but no direct sell order ID listed.. but sounds like it wont work for orders with direct SellOrderId's as with those you either fill it or you don't. 

Is there any spec on how the matching engine will work in this scheme for floating buy/sell orders?

File: x/ecocredit/msgs.go [hunk 6]
```diff
@@ -375,3 +377,156 @@ func (m *MsgUpdateClassMetadata) GetSigners() []sdk.AccAddress {
...
+// Type implements the LegacyMsg interface.
+func (m MsgUpdateSellOrders) Type() string { return sdk.MsgTypeURL(&m) }
+
+// GetSignBytes implements the LegacyMsg interface.
+func (m MsgUpdateSellOrders) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&m))
+}
+
+// ValidateBasic does a sanity check on the provided data.
+func (m *MsgUpdateSellOrders) ValidateBasic() error {
```
[the remark this answers is not in the document]
Comment (technicallyty) [reply x/ecocredit/msgs.go:442]: i see, yeah i think the least complex approach would just be to assume its all there. we'd have to add messages inside messages to get granularity here 🤷🏻

File: proto/regen/ecocredit/v1alpha1/events.proto (hunk 3 above)
Comment (aaronc) [reply proto/regen/ecocredit/v1alpha1/events.proto:156]: > how will the buy order be fulfilled at a later point in time?

Just like in regular exchanges, if the buy order is partially filled it will stay open and way for new or updated sell orders. Basically in every processing batch an attempt will be made to match it with sell orders.

We should probably have expiration times for both sell and buy orders btw.

Comment (ryanchristo) [reply proto/regen/ecocredit/v1alpha1/events.proto:156]: I think I misunderstood "filled". Yes, we should have a separate event for when a buy order is filled.

File: x/ecocredit/server/msg_server.go [hunk 7]
```diff
@@ -577,3 +516,352 @@ func (s serverImpl) isCreatorAllowListed(ctx types.Context, allowlist []string,
...
+	return &ecocredit.MsgBuyResponse{BuyOrderIds: buyOrderIds}, nil
+}
+
+// AllowAskDenom adds a new ask denom
+func (s serverImpl) AllowAskDenom(goCtx context.Context, req *ecocredit.MsgAllowAskDenom) (*ecocredit.MsgAllowAskDenomResponse, error) {
+	ctx := types.UnwrapSDKContext(goCtx)
+
+	if req.RootAddress != authtypes.NewModuleAddress(govtypes.ModuleName).String() {
+		return nil, sdkerrors.ErrUnauthorized.Wrap("root address must be governance module address")
+	}
```
Comment (clevinson) [inline x/ecocredit/server/msg_server.go:774]: @aaronc @blushi Can one of you confirm that this is how the root address stuff is supposed to work?

File: x/ecocredit/msgs.go [hunk 8]
```diff
@@ -375,3 +377,156 @@ func (m *MsgUpdateClassMetadata) GetSigners() []sdk.AccAddress {
...
+		}
+
+		_, err := math.NewPositiveDecFromString(m.Orders[i].Quantity)
+		if err != nil {
+			return sdkerrors.Wrapf(err, "quantity must be positive decimal: %s", m.Orders[i].Quantity)
+		}
+
+		m.Orders[i].AskPrice.Validate()
+
+		if !m.Orders[i].AskPrice.Amount.IsPositive() {
```
[the remark this answers is not in the document]
Comment (clevinson) [reply x/ecocredit/msgs.go:416]: ping @dtkong

File: proto/regen/ecocredit/v1alpha1/events.proto (hunk 3 above)
Comment (clevinson) [reply proto/regen/ecocredit/v1alpha1/events.proto:156]: At this point we havn't implemented any of the ORM tables for storing buy orders in state, but sounds like we would need this in order to actually process partial-filling correctly (where we keep a buy order open). If we don't implement that in this PR we could also just add it in a follow-up. When persisting the partially filled buy order to state, how would we store the `selection` ? Would it still reference the original `sell_order_id` ? Afaiu, the sell order would be exhausted and then closed (removed from state), so there wouldn't be any way for a seller to top up their supply on an existing sell order. They would rather create a new sell order.

If we allow for partial fills I'm not really sure how we can keep those buy orders open seeing that they are against sell order which would have already closed..

Comment (ryanchristo) [reply proto/regen/ecocredit/v1alpha1/events.proto]: If we address in a follow-up, it might make sense to include with #623 assuming that we would only need to store a buy order if it was using the filter option (not a direct buy order).

I added a second event for filling buy orders to distinguish from creating buy orders. At the moment the second event seems unnecessary given that only direct buy orders are implemented in this pull request but it makes sense when we implement the filter option. I think this goes in line with #623 as well.

I also added a couple commits to update the remaining quantity if the buy order does not purchase all credits within the sell order and to delete the sell order if there are no remaining credits.

Maybe we should break out a separate issue for adding and handling expiration times..?

**Comment by ryanchristo** [comment]
Ready for another round of reviews.

Aside from remaining items that already have issues, here are the remaining items that I can see for this pull request that could either be addressed here or in follow-up pull requests:

- [ ] support 0 amount sell orders? @dtkong 
- [ ] create table to store buy orders (maybe it would make sense to include this with #623)
- [ ] update buy order after processing (maybe it would make sense to include this with #623)
- [ ] add expiration times to buy and sell orders

Let me know if you think any of these should be addressed here, otherwise I think it might be nice to include them in follow up issues.

**Review by technicallyty — COMMENTED** [review]
regarding https://github.com/regen-network/regen-ledger/pull/613#issuecomment-965685378, i think it would make sense to include the buy order table here, just to set it up for the processing PR

**Review by technicallyty — APPROVED** [review]
great work! 💯 💯 💯 💯 💯 💯 💯

File: x/ecocredit/server/msg_server.go [hunk 9]
```diff
@@ -577,3 +516,346 @@ func (s serverImpl) isCreatorAllowListed(ctx types.Context, allowlist []string,
 	}
 	return false
 }
+
+// Sell creates new sell orders for credits
+func (s serverImpl) Sell(goCtx context.Context, req *ecocredit.MsgSell) (*ecocredit.MsgSellResponse, error) {
```
[the remark this answers is not in the document]
Comment (aaronc) [reply x/ecocredit/server/msg_server.go:521]: Well maybe we should revisit the pros and cons and escrow. Can you guys think a little more about the pros and cons and the threat model here?

**Review by clevinson — COMMENTED** [review]
Few small nits, but happy to have those covered in followups.

ACK 🎉 -- @ryanchristo can you give an approval for me, since I authored the original PR ?

File: proto/regen/ecocredit/v1alpha1/events.proto [hunk 10]
```diff
@@ -103,3 +105,114 @@ message EventCancel {
...
+  uint64 sell_order_id = 2;
+
+  // batch_denom is the credit batch ID of the purchased credits.
+  string batch_denom = 3;
+
+  // quantity is the quantity of the purchased credits.
+  string quantity = 4;
+
+  // total_price is the total price for the purchased credits.
+  cosmos.base.v1beta1.Coin total_price = 5;
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply proto/regen/ecocredit/v1alpha1/events.proto:203]: The price here is the amount of coin sent (not the price per unit but the calculated amount of coin needed to fill the order). I thought it might help to use total to distinguish from bid price and ask price (per unit) more clearly.

**Review by ryanchristo — APPROVED** [review]
> @ryanchristo can you give an approval for me, since I authored the original PR ?
